tree_test.go

```diff
@@ -1649,3 +1654,127 @@ func TestLeafNodeInsert(t *testing.T) {
 		t.Fatalf("hash should not be nil")
 	}
 }
+
+type randTest []randTestStep
+
+type randTestStep struct {
+	op    int
+	key   []byte // for opInsert, opDelete, opGet
+	value []byte // for opInsert
+	err   error  // for debugging
+}
+
+const (
+	opInsert = iota
+	opDelete
+	opGet
+	opHash
+	opCommit
+	opProve
+	opMax
+)
+
+func (randTest) Generate(r *mRand.Rand, size int) reflect.Value {
+	var finishedFn = func() bool {
+		size--
+		return size == 0
+	}
+	return reflect.ValueOf(generateSteps(finishedFn, r))
+}
+
+func generateSteps(finished func() bool, r io.Reader) randTest {
+	var allKeys [][]byte
+	var one = make([]byte, 32)
+	genKey := func() []byte {
+		r.Read(one)
+		if len(allKeys) < 2 || one[0]%100 > 90 {
+			// new key
+			key := make([]byte, 32)
+			r.Read(key)
+			allKeys = append(allKeys, key)
+			return key
+		}
+		// use existing key
+		idx := int(one[0]) % len(allKeys)
+		return allKeys[idx]
+	}
+	var steps randTest
+	for !finished() {
+		r.Read(one)
+		step := randTestStep{op: int(one[0]) % opMax}
+		switch step.op {
+		case opInsert:
+			step.key = genKey()
+			step.value = make([]byte, 8)
+			binary.BigEndian.PutUint64(step.value, uint64(len(steps)))
+		case opGet, opDelete, opProve:
+			step.key = genKey()
+		}
+		steps = append(steps, step)
+	}
+	return steps
+}
+
+// runRandTestBool coerces error to boolean, for use in quick.Check
+func runRandTestBool(rt randTest) bool {
+	return runRandTest(rt) == nil
+}
+
+func runRandTest(rt randTest) error {
+	var (
+		root   = New()
+		keys   = [][]byte{}
+		values = make(map[string]string)
+		cfg    = GetConfig()
+	)
+	for i, step := range rt {
+		switch step.op {
+		case opInsert:
+			if err := root.Insert(step.key, step.value, nil); err != nil {
+				rt[i].err = err
+			}
+			keys = append(keys, step.key)
+			values[string(step.key)] = string(step.value)
+		case opDelete:
+			if _, err := root.Delete(step.key, nil); err != nil {
+				rt[i].err = err
+			}
+			delete(values, string(step.key))
+		case opGet:
+			v, err := root.Get(step.key, nil)
+			want := values[string(step.key)]
+			if string(v) != want {
+				rt[i].err = fmt.Errorf("mismatch for key %#x, got %#x want %#x, err %v", step.key, v, want, err)
+			}
+		case opProve:
+			if len(keys) == 0 {
+				continue
+			}
+			root.Commit()
+			proof, cis, zis, yis, _ := MakeVerkleMultiProof(root, nil, keys, nil)
+			if ok, err := VerifyVerkleProof(proof, cis, zis, yis, cfg); !ok || err != nil {
+				rt[i].err = fmt.Errorf("could not verify verkle proof: %s, err %v", ToDot(root), err)
+			}
+		case opHash:
+			root.Hash()
+		case opCommit:
+			root.Commit()
```

Comment:
Maybe for these two cases we can check that the returned pointer isn't `nil` since that should never happen.

Makes me feel we should actually not return pointers really. But that prob will have some effects in go-ethereum that we we'd prefer not to handle it now -- maybe we can also add here a `// TODO: reconsider if we should avoid returning pointers in Hash() and Commit()`.